### PR TITLE
Add linear discriminant analysis classifier

### PR DESCRIPTION
**Pull request by RicardoSantos-99**
Adds `Scholar.DiscriminantAnalysis.Linear`, the Linear Discriminant Analysis
classifier. Addresses the LDA item of #338.

LDA fits a Gaussian to each class under a shared within-class covariance,
which gives a linear decision boundary between every pair of classes. The
model exposes `fit/3`, `decision_function/2`, `predict/2` and
`predict_probability/2`, and stores `coefficients`, `intercept`, `means`,
`priors` and `classes`.

A few decisions worth calling out:

* The pooled within-class covariance is normalized by
  `num_samples - num_classes` and the scores are parameterized around the
  overall mean, matching scikit-learn's `svd` solver. That parameterization
  makes `decision_function` differ from the bare textbook formula by a
  class-constant term, which cancels in `predict` and `predict_probability`.
* The coefficients are obtained by solving the pooled covariance system
  directly (`Nx.LinAlg.solve`) rather than via a double SVD. This is
  algebraically equivalent for full-rank data and matches scikit-learn to
  about 1.0e-13, but it assumes an invertible covariance (documented). A
  singular covariance raises instead of silently returning garbage.
* `predict_probability` applies a softmax to the discriminant scores. For
  two classes this reduces exactly to scikit-learn's binary sigmoid.

## Verification

Results match `sklearn.discriminant_analysis.LinearDiscriminantAnalysis`
(solver `svd`) for `coefficients`, `intercept`, `decision_function`,
`predict` and `predict_probability`, on balanced and unbalanced classes,
binary and multiclass, and 2 and 3 features, all to roughly 1.0e-13 in f64.
The reference values in the tests were generated with scikit-learn 1.6.1.

### CODE REVIEW

**Review by josevalim — APPROVED** [review]

File: lib/scholar/discriminant_analysis/linear.ex [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+  Reference:
+
+  * [1] - [The Elements of Statistical Learning, Hastie, Tibshirani & Friedman, Section 4.3](https://hastie.su.domains/ElemStatLearn/)
+  """
+  import Nx.Defn
+  import Scholar.Shared
+
+  @derive {Nx.Container, containers: [:coefficients, :intercept, :means, :priors, :classes]}
+  defstruct [:coefficients, :intercept, :means, :priors, :classes]
```
Comment (krstopro) [inline lib/scholar/discriminant_analysis/linear.ex]: Do you need to store classes given that they are `Nx.iota({num_classes})`? Would it be better to store `num_classes`?

I am not sure about this myself.

File: lib/scholar/discriminant_analysis/linear.ex [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      iex> model = Scholar.DiscriminantAnalysis.Linear.fit(x, y, num_classes: 2)
+      iex> Scholar.DiscriminantAnalysis.Linear.predict(model, Nx.tensor([[-1.5, -1.5], [1.5, 1.5]]))
+      #Nx.Tensor<
+        s32[2]
+        [0, 1]
+      >
+  """
+  defn predict(%__MODULE__{classes: classes} = model, x) do
+    scores = decision_function(model, x)
+    Nx.take(classes, Nx.argmax(scores, axis: 1))
```
Comment (krstopro) [inline lib/scholar/discriminant_analysis/linear.ex]: Do you need `Nx.take/2` here given that your classes are `Nx.iota({num_classes})`?

File: lib/scholar/discriminant_analysis/linear.ex [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+      iex> probs = Scholar.DiscriminantAnalysis.Linear.predict_probability(model, Nx.tensor([[1.5, 1.5]]))
+      iex> Nx.sum(probs) |> Nx.round()
+      #Nx.Tensor<
+        f32
+        1.0
+      >
+  """
+  defn predict_probability(%__MODULE__{} = model, x) do
+    scores = decision_function(model, x)
+    scores = scores - Nx.reduce_max(scores, axes: [1], keep_axes: true)
```
Comment (krstopro) [inline lib/scholar/discriminant_analysis/linear.ex]: You might wanna apply `stop_grad/1` after `Nx.reduce_max/2` as in
https://github.com/elixir-nx/scholar/blob/3619c28c45c760d98c20d81941736f0e7da66183/lib/scholar/linear/logistic_regression.ex#L223

Comment (krstopro) [reply lib/scholar/discriminant_analysis/linear.ex]: Though thinking of it, this function might need to be defined somewhere else, perhaps in `Nx` itself.

**Review by krstopro — COMMENTED** [review]
Few remarks.

File: lib/scholar/discriminant_analysis/linear.ex (hunk 1 above)
Comment (RicardoSantos-99) [reply lib/scholar/discriminant_analysis/linear.ex]: You're right, since the module requires labels to be integers in [0, num_classes), classes is always `Nx.iota({num_classes})` and carries nothing beyond num_classes. Dropping it from the struct. predict then just returns the argmax directly (your next comment).

File: lib/scholar/discriminant_analysis/linear.ex (hunk 2 above)
Comment (RicardoSantos-99) [reply lib/scholar/discriminant_analysis/linear.ex]: Right, `Nx.take(Nx.iota(...), idx)` is just idx. predict now returns `Nx.argmax(scores, axis: 1)` directly.

File: lib/scholar/discriminant_analysis/linear.ex (hunk 3 above)
Comment (RicardoSantos-99) [reply lib/scholar/discriminant_analysis/linear.ex]: Done, wrapped the `reduce_max` in `stop_grad` to match the softmax helpers in logistic regression.

Comment (RicardoSantos-99) [reply lib/scholar/discriminant_analysis/linear.ex]: > Though thinking of it, this function might need to be defined somewhere else, perhaps in `Nx` itself.

Agreed the stable-softmax shows up in a few places now. I left it here for this PR, happy to open a follow-up if you think it belongs in `Scholar.Shared` or upstream in Nx.

**Review by krstopro — APPROVED** [review]
Alright, looks good to me.

@josevalim @polvalente We have softmax several times in `Scholar` and it is implemented in `Axon`. It can be implemented in `Scholar.Shared`, but perhaps it might be better to have it in `Nx`?

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by polvalente** [comment]
@krstopro I think it's ok for us to add softmax in Nx, but we have to think about what types of shapes we want to accept in it (for instance, do we want to always force rank-1 inputs, treat leading axes as batch dims, accept an axes option to calculate the softmax against?). 

Do you wanna submit a PR to Nx with one that can solve both the Axon and Scholar use cases?
I think we should do it as an Nx.Block so that backends can optionally override the implementation.

**Comment by josevalim** [comment]
My .02 cents is that it is a small function that is ok to repeat. The only reason to add it to Nx is only if there are indeed cases where we wnat to override the implementation.

**Comment by polvalente** [comment]
I checked that at least MLX has a fast activation, and we already have sigmoid which I think falls im a similar class of "small and could be copied over" but is still in the Nx API. And softmax is very common overall

**Comment by josevalim** [comment]
So it sounds good to me. And I guess we can look at sigmoid as an example of the API?

**Comment by krstopro** [comment]
> we already have sigmoid which I think falls im a similar class of "small and could be copied over" but is still in the Nx API

Same for `logsumexp` and there are probably others.

> And softmax is very common overall

This was my main argument for it.
